src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,377 @@
+#include "postgres.h"
+#include "libpq-fe.h"
+#include "executor/spi.h"
+#include "distributed/lock_graph.h"
+#include "distributed/coordinator_protocol.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/metadata_utility.h"
+#include "distributed/multi_logical_replication.h"
+#include "postmaster/postmaster.h"
+#include "distributed/distribution_column.h"
+#include "utils/builtins.h"
+#include "distributed/shard_split.h"
+#include "utils/lsyscache.h"
+#include "distributed/listutils.h"
+
+
+PG_FUNCTION_INFO_V1(citus_auto_shard_split_start);
+
+uint64 MaxShardSize = 102400;
+double TenantFrequency = 0.2;
+
+/*
+ * Struct to store all the information related to
+ * a shard.
+ */
+typedef struct ShardInfoData
+{
+	int64 shardSize;
+	int64 shardMinValue;
+	int64 shardMaxValue;
+	int64 shardId;
+	int32 nodeId;
+	char *tableName;
+	char *distributionColumn;
+	char *dataType;
+	char *shardName;
+	Oid tableId;
+	Oid distributionColumnId;
+}ShardInfoData;
+typedef ShardInfoData *ShardInfo;
+
+
+/*
+ * It throws an error if a concurrent automatic shard split or Rebalance operation is happening.
+ */
+void
+ErrorOnConcurrentOperation()
+{
+	int64 jobId = 0;
+	if (HasNonTerminalJobOfType("rebalance", &jobId))
+	{
+		ereport(ERROR, (
+					errmsg("A rebalance is already running as job %ld", jobId),
+					errdetail("A rebalance was already scheduled as background job"),
+					errhint("To monitor progress, run: SELECT * FROM "
+							"citus_rebalance_status();")));
+	}
+	if (HasNonTerminalJobOfType("Automatic Shard Split", &jobId))
+	{
+		ereport(ERROR, (
+					errmsg("An automatic shard split is already running as job %ld",
+						   jobId),
+					errdetail(
+						"An automatic shard split was already scheduled as background job")));
+	}
+}
+
+
+/*
+ * For a given SplitPoints , it creates the SQL query for the Shard Splitting
+ */
+StringInfo
+GetShardSplitQuery(ShardInfo shardinfo, List *splitPoints, char *shardSplitMode)
+{
+	StringInfo splitQuery = makeStringInfo();
+
+	int64 length = list_length(splitPoints);
+	appendStringInfo(splitQuery, "SELECT citus_split_shard_by_split_points(%ld, ARRAY[",
+					 shardinfo->shardId);
+
+	int32 splitpoint = 0;
+	int index = 0;
+	foreach_int(splitpoint, splitPoints)
+	{
+		appendStringInfo(splitQuery, "'%d'", splitpoint);
+
+		if (index < length - 1)
+			appendStringInfoString(splitQuery, ",");
+		
+		index++;
+	}
+
+	appendStringInfo(splitQuery, "], ARRAY[");
+
+	for (int i = 0; i < length; i++)
```

Comment:
would be good to add a comment that we currently write all segments to the same node

Reply:
Comment added